lib/puppet/type/zfs_share.rb

```diff
@@ -0,0 +1,234 @@
+Puppet::Type.newtype(:zfs_share) do
+  @doc = "Manage zfs shares"
+
+  require 'puppet/parameter/boolean'
+
+  ensurable
+
+  # zfs share documentation
+  # https://docs.oracle.com/cd/E26502_01/html/E29031/zfs-share-1m.html
+
+  newparam(:name) do
+    desc "Name of zfs file system"
+    isnamevar
+    validate do |value|
+      if value =~ /^\//
+        raise ArgumentError , "#{value} is not a valid zfs title"
+      end
+    end
+  end
+
+  newparam(:share_title) do
+    desc "Creates a named share with the share title"
+    validate do |value|
+      if value =~ /\s/
+        raise ArgumentError , "#{value} is not a valid share title"
+      end
+    end
+  end
+
+  newproperty(:nfs) do
+    desc "Enable or disable nfs"
+    defaultto 'on'
+    newvalues(:on, :off)
+  end
+
+  newproperty(:nfs_log) do
+    desc "Enables NFSv2 or NFSv3 server logging for the specified file system. The tag is defined in the /etc/nfs/nfslog.conf file. If no tag is specified, the default values associated with the global tag in the /etc/nfs/nfslog.conf file is used."
+  end
+
+  newproperty(:nosub) do
+    desc "Prevents NFSv2 or NFSv3 clients from mounting subdirectories of shared directories."
+    newvalues(:on, :off)
+  end
+
+  newparam(:guestok) do
+    desc "Whether to allow guest access to smb shares"
+    newvalues(:on, :off)
+  end
+
+  newproperty(:sec_none_rw, :array_matching => :all) do
+    validate do |value|
+      unless value =~ /^\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\/\d{1,2}$|\*/
+        raise ArgumentError , "#{value} is not a valid IP address."
```

Comment:
I am not sure validating against an IP address is a good idea.  NFS ACLs can be IPs, hostnames with wildcards, or netgroups.